### ISSUE TITLE
Support parameterized views (efficiently)

### ISSUE THREAD

**Issue by quodlibetor** [MaterializeInc/materialize#2210]
If you want to add a `WHERE` condition to a materialized view, the only option right now is to create a view and then `PREPARE` a query:

```sql
CREATE MATERIALIZED VIEW v AS VALUES (1), (2), (3), (4);
-- currently we only support the prepare via the pgwire codec, but conceptually:
PREPARE q (int) AS SELECT * FROM v WHERE column1 > $1;
```

This has a couple primary drawbacks:

* It's not particularly performant: I believe that we need to create a new
  dataflow every time a parameter is supplied
* It's not particularly ergonomic: it's not obvious that this is a possible set
  of steps
* It conflates ad-hoc occasional queries with things that want to be done for
  performance: a parameterized view gives us information that the user expects
  this data to be maintained, which suggests that we should add an index on the
  parameters

I'm not sure what the best syntax for this would be. AFAICT no databases actually support parameterized views, but all databases have very different semantics to us.

As a strawman I'd propose following the `PREPARE` syntax and just allowing data types and parameters, with the goal of providing automatic indexing for fast queries:

```sql
CREATE PARAMETERIZED MATERIALIZED VIEW v (int) AS
SELECT * FROM source WHERE source.c > $1
```

**Comment by quodlibetor** [comment]
An important motivation for this IMO is supporting things like `query_now()` functions, which can desugar down to parameterized views that take in an implicit parameter from the coordinator.

**Comment by benesch** [comment]
> I'm not sure what the best syntax for this would be.

The best syntax is the syntax that we already support! The syntax you included at the start of the issue, that is:

```sql
CREATE MATERIALIZED VIEW v AS VALUES (1), (2), (3), (4);
-- currently we only support the prepare via the pgwire codec, but conceptually:
PREPARE q (int) AS SELECT * FROM v WHERE column1 > $1;
```

> It's not particularly performant: I believe that we need to create a new
> dataflow every time a parameter is supplied

Yes, but that's an implementation detail. We just need to write the code that applies the filter to the underlying arrangement without constructing a new dataflow. All of the information that we need to make that determination is available already.

> It's not particularly ergonomic: it's not obvious that this is a possible set
of steps

I disagree. This is exactly the set of steps that you would use in any other database.

> It conflates ad-hoc occasional queries with things that want to be done for
> performance: a parameterized view gives us information that the user expects
> this data to be maintained, which suggests that we should add an index on the
> parameters

Automatic index selection is an essentially unsolved problem. We might be able to do a decent job on the very simple cases, but the complexity quickly spirals. I think we are much better off in the near futures asking users to carefully consider what indexes to build, just like you'd have to in a traditional RDMBS. 

tl;dr I really think we just need to add some implementation smarts to optimize the case where you are filtering an existing arrangement. I think introducing new syntax here will just confuse users!

**Comment by quodlibetor** [comment]
We don't currently support `PREPARE` in sql, although that is just an engineering issue.

> We just need to write the code that applies the filter to the underlying arrangement without constructing a new dataflow.

Yeah, this issue is stating that we should do that. I should have been more clear that fixing prepare is equivalent to supporting parameterized views, and that I felt like there should be a distinction between prepared statements and parameterized views. Although that distinction could just be "did you create an index before you defined the view".

> Automatic index selection is an essentially unsolved problem. We might be able to do a decent job on the very simple cases, but the complexity quickly spirals. I think we are much better off in the near futures asking users to carefully consider what indexes to build, just like you'd have to in a traditional RDMBS.

That's fair, but also a huge amount of what we're doing already is automatic index creation, so while I agree that the `CREATE VIEW`/`PREPARE q`/`CREATE INDEX` is a totally reasonable (and certainly initially the right) way to do this, it's not out of the question that we're in a much better position to automatically create indexes than other RDBMSs.

**Comment by benesch** [comment]
Oh yeah, sorry, should have been more clear that I’m definitely on board
with this issue existing to track that improvement!

**Comment by benesch** [comment]
Revisiting this after 2+ years: the idea of creating indexes automatically on `PREPARE` stresses me out more now than it did then. The performance characteristics seem difficult to reason about. Mostly because we have to figure out when to _drop_ those indexes. Unlike with `CREATE INDEX`, which is paired with an explicit `DROP INDEX` command. The inverse of `PREPARE` is `DEALLOCATE` (or closing the session), but.. common usage patterns will involve multiple connections from the same application all `PREPARE`ing the same statement. We'd need to share and refcount the index across connections... but then you have a cold start problem. If for some reason all the connections drop, then the index drops, and the next connection will have to build a new dataflow from scratch.

I think a perhaps better direction to go is an index suggestion wizard! You type your query in to Materialize, and Materialize suggests some indexes to make! Maybe `EXPLAIN` could do this: make some suggestions about a view/index you could make to speed up the query you just explained.

**Comment by quodlibetor** [comment]
This issue also very much predates all the UX we built around `TAIL`/`SUBSCRIBE`. I believe this was just one of the brainstorming ideas that we had along the way to SUBSCRIBE, and that SUBSCRIBE (possibly with some index recommendation UX) is probably the right resolution to the problem this is trying to solve.

I also agree that automatic index creation and deletion would be scary.
